File: libs/idun_agent_schema/src/idun_agent_schema/engine/graph.py
"""Framework-agnostic graph IR shared by engine and UI consumers.

Versioned via `format_version`. New variants ship alongside, never as a
breaking change.
"""

from __future__ import annotations

from enum import Enum
from typing import Annotated, Literal

from pydantic import BaseModel, Field

from idun_agent_schema.engine.agent_framework import AgentFramework


class AgentKind(str, Enum):
    LLM = "llm"
    SEQUENTIAL = "sequential"
    PARALLEL = "parallel"
    LOOP = "loop"
    CUSTOM = "custom"


class ToolKind(str, Enum):
    NATIVE = "native"
    MCP = "mcp"
    BUILT_IN = "built_in"


class EdgeKind(str, Enum):
    PARENT_CHILD = "parent_child"
    SEQUENTIAL_STEP = "sequential_step"
    PARALLEL_BRANCH = "parallel_branch"
    LOOP_STEP = "loop_step"
    TOOL_ATTACH = "tool_attach"
    GRAPH_EDGE = "graph_edge"


class AgentNode(BaseModel):
    kind: Literal["agent"] = "agent"
    id: str
    name: str
    agent_kind: AgentKind
    is_root: bool = False
    description: str | None = None
    model: str | None = None
    loop_max_iterations: int | None = None


class ToolNode(BaseModel):
    kind: Literal["tool"] = "tool"
    id: str
    name: str
    tool_kind: ToolKind
    description: str | None = None
    mcp_server_name: str | None = None


AgentGraphNode = Annotated[
    AgentNode | ToolNode, Field(discriminator="kind")
]


class AgentGraphEdge(BaseModel):
    source: str
    target: str
    kind: EdgeKind
    order: int | None = None
    condition: str | None = None
    label: str | None = None


class AgentGraphMetadata(BaseModel):
    framework: AgentFramework
    agent_name: str
    root_id: str
    warnings: list[str] = Field(default_factory=list)


class AgentGraph(BaseModel):
    format_version: Literal["1"] = "1"
    metadata: AgentGraphMetadata
    nodes: list[AgentGraphNode]
    edges: list[AgentGraphEdge]

File: services/idun_agent_standalone_ui/lib/api/types/graph.ts
/**
 * TypeScript mirror of idun_agent_schema.engine.graph (Pydantic).
 *
 * The schema package is the source of truth — when those models change, this
 * file must follow. Discriminated unions on `kind`.
 */

export type AgentKind =
  | "llm"
  | "sequential"
  | "parallel"
  | "loop"
  | "custom";

export type ToolKind = "native" | "mcp" | "built_in";

export type EdgeKind =
  | "parent_child"
  | "sequential_step"
  | "parallel_branch"
  | "loop_step"
  | "tool_attach"
  | "graph_edge";

// `extends Record<string, unknown>` satisfies @xyflow/react v12's
// `Node<TData extends Record<string, unknown>>` / `Edge<TData ...>`
// generic constraint when these types are used as ReactFlow payloads.
// Doesn't change the wire shape — Pydantic emits the same JSON.

export interface AgentNode extends Record<string, unknown> {
  kind: "agent";
  id: string;
  name: string;
  agent_kind: AgentKind;
  is_root: boolean;
  description: string | null;
  model: string | null;
  loop_max_iterations: number | null;
}

export interface ToolNode extends Record<string, unknown> {
  kind: "tool";
  id: string;
  name: string;
  tool_kind: ToolKind;
  description: string | null;
  mcp_server_name: string | null;
}

export type AgentGraphNode = AgentNode | ToolNode;

export interface AgentGraphEdge extends Record<string, unknown> {
  source: string;
  target: string;
  kind: EdgeKind;
  order: number | null;
  condition: string | null;
  label: string | null;
}

export interface AgentGraphMetadata {
  framework: "LANGGRAPH" | "ADK" | "HAYSTACK";
  agent_name: string;
  root_id: string;
  warnings: string[];
}

export interface AgentGraph {
  format_version: "1";
  metadata: AgentGraphMetadata;
  nodes: AgentGraphNode[];
  edges: AgentGraphEdge[];
}
